### PR TITLE
Soft deletes support in BlobStoreRepository and using it for Remote Store

### PR DESCRIPTION
**Pull request by gbbafna**
### Description
1. Introduces Soft Delete Support on Repository Interface and corresponding S3 Repository changes to support it 

1. Using Soft Deletes in Remote Store 

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Pending 
1. Performance tests
2. Failing UT fix 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/common/blobstore/SoftDeleteSupportedBlobContainer.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * An extension of {@link BlobContainer} that adds support to soft delete files
+ *
+ * @opensearch.internal
+ */
+public interface SoftDeleteSupportedBlobContainer {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/common/blobstore/SoftDeleteSupportedBlobContainer.java:19]: Rename `SoftDeleteableBlobContainer`

File: server/src/main/java/org/opensearch/common/blobstore/SoftDeleteSupportedBlobContainer.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    /**
+     * Soft Deletes the blobs with given names. This method will not throw an exception
+     * when one or multiple of the given blobs don't exist and simply ignore this case.
+     *
+     * @param   blobNames  The names of the blob to delete.
+     * @throws  IOException if a subset of blob exists but could not be soft deleted.
+     */
+    void softDeleteBlobsIgnoringIfNotExists(List<String> blobNames) throws IOException;
+
+    DeleteResult  softDelete() throws IOException;
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/common/blobstore/SoftDeleteSupportedBlobContainer.java:29]: nit: extra space

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/common/blobstore/SoftDeleteSupportedBlobContainer.java:29]: Also add java docs

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobStore.java [hunk 3]
```diff
@@ -76,6 +79,10 @@ class S3BlobStore implements BlobStore {
     private final AsyncExecutorContainer normalExecutorBuilder;
     private final boolean multipartUploadEnabled;
 
+    final Setting<Boolean> SOFT_DELETE_ENABLED = Setting.boolSetting("soft_delete.enabled", true);
```
Comment (Bukhtawar) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobStore.java:82]: do we need to add Node.Scope

File: server/src/main/java/org/opensearch/index/translog/transfer/BlobStoreTransferService.java [hunk 4]
```diff
@@ -160,7 +161,12 @@ public InputStream downloadBlob(Iterable<String> path, String fileName) throws I
 
     @Override
     public void deleteBlobs(Iterable<String> path, List<String> fileNames) throws IOException {
-        blobStore.blobContainer((BlobPath) path).deleteBlobsIgnoringIfNotExists(fileNames);
+        if (blobStore.blobContainer((BlobPath) path) instanceof SoftDeleteSupportedBlobContainer) {
+            ((SoftDeleteSupportedBlobContainer) blobStore.blobContainer((BlobPath) path)).softDeleteBlobsIgnoringIfNotExists(fileNames);
+            logger.info("Yay soft deelted");
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/translog/transfer/BlobStoreTransferService.java:166]: We will be changing this, right? Also, typo in `deleted` :)

File: server/src/main/java/org/opensearch/common/blobstore/SoftDeleteSupportedBlobContainer.java (hunk 1 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/common/blobstore/SoftDeleteSupportedBlobContainer.java:19]: done.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobStore.java (hunk 3 above)
Comment (gbbafna) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobStore.java:82]: this is related to repo metadata only . So not added it . Same is followed for other repo settings as well .

File: server/src/main/java/org/opensearch/index/translog/transfer/BlobStoreTransferService.java (hunk 4 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/translog/transfer/BlobStoreTransferService.java]: yes, this is just debug logs. removed it now .

File: server/src/main/java/org/opensearch/common/blobstore/SoftDeleteableSupportedBlobContainer.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * An extension of {@link BlobContainer} that adds support to soft delete files
+ *
+ * @opensearch.internal
+ */
+public interface SoftDeleteableSupportedBlobContainer {
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/common/blobstore/SoftDeleteableSupportedBlobContainer.java:19]: can we make this to just `SoftDeleteableBlobContainer`? Supported is implicit given implementor needs to implement this interface.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 6]
```diff
@@ -248,11 +249,16 @@ public DeleteResult delete() throws IOException {
                     return s3Object.key();
                 }).collect(Collectors.toList());
 
-                if (!listObjectsResponseIterator.hasNext()) {
+                if (!listObjectsResponseIterator.hasNext() && softDelete == false) {
```
Comment (ashking94) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:252]: Why are we not marking to delete the key path when softDelete is true?

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobStore.java [hunk 7]
```diff
@@ -186,6 +193,18 @@ public static StorageClass initStorageClass(String storageClassStringValue) {
...
+    public Tagging softDeleteTagging() {
+        Tag tag = Tag.builder()
+            .key(SOFT_DELETE_KEY.get(repositoryMetadata.settings()))
+            .value(SOFT_DELETE_VALUE.get(repositoryMetadata.settings()))
+            .build();
+        return Tagging.builder().tagSet(tag).build();
+    }
+
+    boolean softDeleteEnabled() {
+        return SOFT_DELETE_ENABLED.get(repositoryMetadata.settings());
```
Comment (ashking94) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobStore.java:205]: Just curious on why we are relying on a setting rather than a overridden method that can supply is with this information whether the soft deletion capability exists or not. With this we need to handle so many combinations of the actual support and passed settings which can probably overcomplicate things and will require us to handle them in code without a real need to do so.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 8]
```diff
@@ -626,4 +637,53 @@ static Tuple<Long, Long> numberOfMultiparts(final long totalSize, final long par
...
+            for (String key : outstanding) {
+                SocketAccess.doPrivilegedVoid(() -> {
+                    PutObjectTaggingRequest putObjectTaggingRequest = PutObjectTaggingRequest.builder()
+                        .bucket(blobStore.bucket())
+                        .key(key)
+                        .tagging(blobStore.softDeleteTagging())
+                        .build();
+                    clientReference.get().putObjectTagging(putObjectTaggingRequest);
+                });
+            }
```
Comment (ashking94) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:676]: Could this rather be exposed as a feature to add tags? For remote store, we can access this method for adding tags for deletion. Since there is an additional step of creation a rule or a policy, it is not an soft deletion feature offered by the remote store, but us using some other remote store features to implement soft deletion.

File: server/src/main/java/org/opensearch/common/blobstore/BlobStore.java [hunk 9]
```diff
@@ -53,4 +53,5 @@ public interface BlobStore extends Closeable {
     default Map<String, Long> stats() {
         return Collections.emptyMap();
     }
+
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/common/blobstore/BlobStore.java:56]: nit - remove this unnecessary change

**Review by ashking94 — COMMENTED** [review]
Left some comments.

File: server/src/main/java/org/opensearch/common/blobstore/BlobStore.java (hunk 9 above)
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/common/blobstore/BlobStore.java:56]: nit: Added by mistake?

File: server/src/main/java/org/opensearch/index/store/RemoteDirectory.java [hunk 10]
```diff
@@ -119,7 +120,11 @@ public void onFailure(Exception e) {
     @Override
     public void deleteFile(String name) throws IOException {
         // ToDo: Add a check for file existence
-        blobContainer.deleteBlobsIgnoringIfNotExists(Collections.singletonList(name));
+        if (blobContainer instanceof SoftDeleteableSupportedBlobContainer) {
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/store/RemoteDirectory.java:123]: Should caller be aware of this or can be re-directed as part of `blobContainer` method?

**Comment by sachinpkale** [comment]
Please add PR description and resolve conflicts.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 6 above)
Comment (gbbafna) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:252]: we can't delete the path when the objects in it are still existing .

File: server/src/main/java/org/opensearch/common/blobstore/SoftDeleteableSupportedBlobContainer.java (hunk 5 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/common/blobstore/SoftDeleteableSupportedBlobContainer.java]: removed interface now .

**Comment by gbbafna** [comment]
Closing it for now. We will revisit it once we are ready to support it comprehensively .
